src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java

```diff
@@ -0,0 +1,249 @@
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
+            throws IOException {
+        Map<String, List<String>> headers = new HashMap<>();
+        headers.put(
+                "Set-Cookie", Arrays.asList("CSRF_Level_3=" + FIXED_TOKEN + "; HttpOnly; Secure"));
+        return new ResponseEntity<GenericVulnerabilityResponseBean<String>>(
+                new GenericVulnerabilityResponseBean<String>(
+                        checkInputsToken(usrLiveToken, FIXED_TOKEN, usrId, usrPass, usrNewPass),
+                        true),
+                CollectionUtils.toMultiValueMap(headers),
+                HttpStatus.OK);
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.CSRF,
+            description = "CSRF_VULNERABILITY_USER_INPUT1",
+            payload = "CSRF_PAYLOAD_LEVEL_3")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_4,
+            htmlTemplate = "LEVEL_4/CSRFVulnerability",
+            requestMethod = RequestMethod.POST)
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl4(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
+            throws IOException {
+        Map<String, List<String>> headers = new HashMap<>();
+        headers.put(
+                "Set-Cookie", Arrays.asList("CSRF_Level_4=" + FIXED_TOKEN + "; HttpOnly; Secure"));
+        return new ResponseEntity<GenericVulnerabilityResponseBean<String>>(
+                new GenericVulnerabilityResponseBean<String>(
+                        checkInputsToken(usrLiveToken, FIXED_TOKEN, usrId, usrPass, usrNewPass),
+                        true),
+                CollectionUtils.toMultiValueMap(headers),
+                HttpStatus.OK);
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.CSRF,
+            description = "CSRF_VULNERABILITY_USER_INPUT1",
+            payload = "CSRF_PAYLOAD_LEVEL_5")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_5,
+            htmlTemplate = "LEVEL_5/CSRFVulnerability",
+            variant = Variant.SECURE)
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl5(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
+            throws IOException {
+        Map<String, List<String>> headers = new HashMap<>();
+        if (usrLiveToken.equals("secretrequest")) {
```

Comment:
what is the logic of this if check?

Reply:
I think we can just read the cookie here and see if secret token cookie is there or not and if it there then allow the change of password, else return to login screen.

Reply:
I am not sure about this, it was done by the original author of the file. I'll dig into it